controller/lib/core/comms.cpp

```diff
@@ -204,54 +37,77 @@ void comms_sendControllerStatus(ControllerStatus controller_status) {
 // TODO add CRC to whole packet
 
 // TODO run this via DMA to free up resources for control loops
-static void process_tx() {
-  if (output_buffer_ready) {
-    for (uint16_t i = 0; i < tx_data_length; i++) {
-      uint16_t written = Hal.serialWrite(tx_buffer[i]);
-      if (1 != written) {
-        // TODO catch on fire
-      }
+static void process_tx(const ControllerStatus &controller_status) {
+  auto bytes_avail = Hal.serialBytesAvailableForWrite();
+
+  // Serialize our current state into the buffer if
+  //  - we're not currently transmitting,
+  //  - we can transmit at least one byte now, and
+  //  - it's been a while since we last transmitted.
+  if (tx_bytes_remaining == 0 && bytes_avail > 0 &&
+      (last_tx_ms == static_cast<decltype(last_tx_ms)>(-1) ||
+       Hal.millis() - last_tx_ms > TX_TIMEOUT_MS)) {
+    // Serialize current status into output buffer.
+    //
+    // TODO: Frame the message bytes.
+    // TODO: Add a checksum to the message.
+    pb_ostream_t stream = pb_ostream_from_buffer(tx_buffer, sizeof(tx_buffer));
+    if (!pb_encode(&stream, ControllerStatus_fields, &controller_status)) {
+      // TODO: Serialization failure; log an error or raise an alert.
+      return;
     }
-    output_buffer_ready = false;
+    tx_idx = 0;
+    tx_bytes_remaining = stream.bytes_written;
+    last_tx_ms = Hal.millis();
+  }
+
+  // TODO: Alarm if we haven't been able to send a status in a certain amount
+  // of time.
+
+  // Send bytes over the wire if any are in our buffer.
+  if (tx_bytes_remaining > 0 && bytes_avail > 0) {
+    auto bytes_written =
+        Hal.serialWrite(reinterpret_cast<char *>(tx_buffer) + tx_idx,
```

Comment:
Would it be possible to have serialWrite take the same thing that nanopb wants? (uint8?)

Reply:
Left a TODO; this patch is already bigger than I like.  Hope that's OK.